subprojects/common/core/src/testFixtures/java/hu/bme/mit/theta/core/utils/FpTestUtils.java

```diff
@@ -0,0 +1,181 @@
+				{FpRoundToIntegralExpr.class, Fp16("2.0"), RoundToIntegral(RNE, Fp16("2.49"))},
+				{FpRoundToIntegralExpr.class, Fp16("-10.0"), RoundToIntegral(RNE, Fp16("-10.49"))},
+				{FpSqrtExpr.class, Fp16("2.1"), Sqrt(RNE, Fp16("4.41"))},
+				{FpSqrtExpr.class, Fp16("3.0"), Sqrt(RNE, Fp16("9.0"))},
+				{FpSqrtExpr.class, Fp16("0.1"), Sqrt(RNE, Fp16("0.01"))},
+				{FpToBvExpr.class, BvUtils.bigIntegerToUnsignedBvLitExpr(BigInteger.TEN, 16), ToBv(RTZ, Fp16("10.9"), 16, false)},
+				{FpToBvExpr.class, BvUtils.bigIntegerToUnsignedBvLitExpr(BigInteger.TEN, 3), ToBv(RTZ, Fp16("10.9"), 3, false)},
+				{FpToBvExpr.class, BvUtils.bigIntegerToSignedBvLitExpr(BigInteger.TEN, 16), ToBv(RTZ, Fp16("10.9"), 16, true)},
+				{FpToBvExpr.class, BvUtils.bigIntegerToSignedBvLitExpr(BigInteger.TEN, 4), ToBv(RTZ, Fp16("10.9"), 4, true)},
+				{FpToBvExpr.class, BvUtils.bigIntegerToSignedBvLitExpr(BigInteger.TEN.negate(), 16), ToBv(RTZ, Fp16("-10.9"), 16, true)},
+				{FpToBvExpr.class, BvUtils.bigIntegerToSignedBvLitExpr(BigInteger.TEN.negate(), 4), ToBv(RTZ, Fp16("-10.9"), 4, true)},
+				{FpToFpExpr.class, Fp32("12.0"), ToFp(RNE, Fp16("12.0"), 8, 24)},
+				{FpToFpExpr.class, Fp64("12.0"), ToFp(RNE, Fp16("12.0"), 11, 53)},
+				{FpToFpExpr.class, Fp16("12.0"), ToFp(RNE, Fp32("12.0"), 5, 11)},
+				{FpFromBvExpr.class, Fp16("0"), FromBv(RNE, BvUtils.bigIntegerToUnsignedBvLitExpr(BigInteger.ZERO, 16), FpType(5, 11), false)},
+				{FpFromBvExpr.class, Fp16("1"), FromBv(RNE, BvUtils.bigIntegerToUnsignedBvLitExpr(BigInteger.ONE, 16), FpType(5, 11), false)},
+				{FpFromBvExpr.class, Fp16("10"), FromBv(RNE, BvUtils.bigIntegerToUnsignedBvLitExpr(BigInteger.TEN, 16), FpType(5, 11), false)},
+				{FpFromBvExpr.class, Fp16("0"), FromBv(RNE, BvUtils.bigIntegerToSignedBvLitExpr(BigInteger.ZERO, 16), FpType(5, 11), true)},
+				{FpFromBvExpr.class, Fp16("1"), FromBv(RNE, BvUtils.bigIntegerToSignedBvLitExpr(BigInteger.ONE, 16), FpType(5, 11), true)},
+				{FpFromBvExpr.class, Fp16("10"), FromBv(RNE, BvUtils.bigIntegerToSignedBvLitExpr(BigInteger.TEN, 16), FpType(5, 11), true)},
+				{FpFromBvExpr.class, Fp16("0"), FromBv(RNE, BvUtils.bigIntegerToSignedBvLitExpr(BigInteger.ZERO.negate(), 16), FpType(5, 11), true)},
+				{FpFromBvExpr.class, Fp16("-1"), FromBv(RNE, BvUtils.bigIntegerToSignedBvLitExpr(BigInteger.ONE.negate(), 16), FpType(5, 11), true)},
+				{FpFromBvExpr.class, Fp16("-10"), FromBv(RNE, BvUtils.bigIntegerToSignedBvLitExpr(BigInteger.TEN.negate(), 16), FpType(5, 11), true)},
+		});
+	}
+
+	public static Collection<?> NaNOperations() {
+		return Arrays.asList(new Object[][]{
+				{FpAddExpr.class, Fp16NaN(), Add(RNE, List.of(Fp16NaN(), Fp16("3.4")))},
+				{FpSubExpr.class, Fp16NaN(), Sub(RNE, Fp16("5.5"), Fp16NaN())},
+				{FpPosExpr.class, Fp16NaN(), Pos(Fp16NaN())},
+				{FpNegExpr.class, Fp16NaN(), Neg(Fp16NaN())},
+				{FpMulExpr.class, Fp16NaN(), Mul(RNE, List.of(Fp16NaN(), Fp16("3.4")))},
+				{FpDivExpr.class, Fp16NaN(), Div(RNE, Fp16("7.14"), Fp16NaN())},
```

Comment:
I think we should write some tests to test the relational operations with NaN values. They are quite the pain in the *** according to the standard.

Reply:
Added some tests in [9733f6f](https://github.com/ftsrg/theta/pull/123/commits/9733f6fc2817b46f5078d5c11c1d6db8b2c35ceb) - are these what you had in mind?

Reply:
Yes, exactly!